### PR TITLE
Cluster-autoscaler: cluster state registry proposal

### PR DESCRIPTION
**Pull request by mwielgus**
The proposal that addresses #2228.

cc: @fgrzadkowski @jszczepkowski @piosz @mumoshu @andrewsykim @kubernetes/sig-autoscaling-misc

### CODE REVIEW

File: cluster-autoscaler/proposals/clusterstate.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+* [UC2] A new node is being added to the cluster. The node has registered on the cluster but has not yet switched its state to ready. This should be fixed in couple seconds. Indicating factors:
+ -- The unready node is new. CreateTime in the last couple minutes.
+Suggested action: Continue operations, however include all yet-to-arrive nodes in all scale-up considerations.
+
+* [UC3] A new node was added to the cluster but failed to start within the reasonable time. There is little chance that it will start anytime soon. Indicating factors:
+ -- Node is unready
+ -- CreateTime == unready NodeCondition.LastTransitionTime
+Suggested action: Continue operations, however do not expand this node pool. The probable scenario is that the node will be picked by scale down soon.
+
+* [UC3] A new node is being added to the cluster. However the cloud provider cannot provision the node within the reasonable time due to either no quota or technical problems. Indicating factors:
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: UC3 appears twice

File: cluster-autoscaler/proposals/clusterstate.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+### Cluster State Registry
+### Handling unready nodes 
+
+### Introduction
+
+Currently ClusterAutoscaler stops working when the number of nodes observed on the cloud provider side doesn’t match to the number of ready nodes on the Kubernetes side. This behavior was introduced in the first version of CA in order to prevent CA from breaking more the already broken cluster. However the client reported feedback indicated that this behavior is oftentimes suboptimal and leads to confusion. This document sketches how the problem of unready nodes will be solved in the next release of K8S.
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: s/the already broken cluster/an already broken cluster
s/the client/clients

File: cluster-autoscaler/proposals/clusterstate.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+Suggested action: Continue operations, however include all yet-to-arrive nodes in all scale-up considerations.
+
+* [UC2] A new node is being added to the cluster. The node has registered on the cluster but has not yet switched its state to ready. This should be fixed in couple seconds. Indicating factors:
+ -- The unready node is new. CreateTime in the last couple minutes.
+Suggested action: Continue operations, however include all yet-to-arrive nodes in all scale-up considerations.
+
+* [UC3] A new node was added to the cluster but failed to start within the reasonable time. There is little chance that it will start anytime soon. Indicating factors:
+ -- Node is unready
+ -- CreateTime == unready NodeCondition.LastTransitionTime
+Suggested action: Continue operations, however do not expand this node pool. The probable scenario is that the node will be picked by scale down soon.
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: Shouldn't w add a condition here that the node is unready for a long enough time?

File: cluster-autoscaler/proposals/clusterstate.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+* Is the cluster, in general, in a good enough shape for CA to operate. The cluster is in the good shape if most of the nodes are in the ready state, and the number of nodes that are in the unjustified unready state (not related to scale down or scale up operations) is limited. CA should halt operations if the cluster is unhealthy and alert the system administrator.
+
+* Is the given Node group, in general, in a good enough shape for CA to operate on it. The NodeGroup is in the good shape if the number of nodes that are unready (but not due to current scale-up/scale-down operations) or not present at all (not yet started by cloud provider) is limited. CA should take extra care about these and not scale up them further until the situation improves. 
+
+* What nodes should soon arrive to the cluster. So that estimator takes them into account and don't ask again for resources for the already handled pods. Also, with that, the estimator won't need to wait for nodes to appear in the cluster.
+
+* How long the given node group was missing nodes. If a fixed number of nodes is missing for a long time this may indicate quota problems. Such node groups should be resized to the actual size.
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: s/was/has

File: cluster-autoscaler/proposals/clusterstate.md (hunk 3 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/clusterstate.md]: Done.

File: cluster-autoscaler/proposals/clusterstate.md (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/clusterstate.md]: Done.

File: cluster-autoscaler/proposals/clusterstate.md (hunk 4 above)
Comment (mwielgus) [inline cluster-autoscaler/proposals/clusterstate.md]: Done.

File: cluster-autoscaler/proposals/clusterstate.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+* [S2] Is the given Node group, in general, in a good enough shape for CA to operate on it. The NodeGroup is in the good shape if the number of nodes that are unready (but not due to current scale-up/scale-down operations) or not present at all (not yet started by cloud provider) is limited. CA should take extra care about these and not scale up them further until the situation improves. 
+
+* [S3] What nodes should soon arrive to the cluster. So that estimator takes them into account and don't ask again for resources for the already handled pods. Also, with that, the estimator won't need to wait for nodes to appear in the cluster.
+
+* [S4] How long the given node group have been missing nodes. If a fixed number of nodes is missing for a long time this may indicate quota problems. Such node groups should be resized to the actual size.
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: s/have/has

File: cluster-autoscaler/proposals/clusterstate.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+### Algorithm
+
+The main loop algorithm will look as follows:
+
+1. Get all nodes from the cluster.
+2. Check if a cluster is, in general, in good health. If not, alert the user, skip the iteration and wait 10 seconds. Helps with [UC7], uses [S1].
+3. Check if any of the node groups has long-time missing nodes. If yes, reduce the size of the node group by the number
+of long-missing nodes. Skip the rest of the iteration. Helps with [UC4], uses [S4].
+4. Check if there are any pending pods. Skip these that can be scheduled on the currently available ready nodes, skipping these that are to be deleted [UC6]. 
+5. If there are still some pending pods,find which of the node group can be expanded to accommodate them. Skip node groups that are not healthy. Helps with [UC3], uses [S2].
```
Comment (jszczepkowski) [inline cluster-autoscaler/proposals/clusterstate.md]: s/pods,find/pods, find/

**Comment by jszczepkowski** [comment]
LGTM
(after off-line discussion)

**Pull request merged**